File: python/b12865.py

# 12865: 평범한 배낭

def solution(N, K):
	for i in range(1, K+1):
		for j in range(1, N+1):
			if w[j] > i:
				dp[j][i] = dp[j-1][i]
			else:
				dp[j][i] = max(dp[j-1][i-w[j]] + v[j], dp[j-1][i])

inputs = input().split(" ")
N = int(inputs[0])
K = int(inputs[1])

dp = list()
for i in range(N+1):
	d = list()
	for j in range(K+1):
		d.append(0)
	dp.append(d)

w = list()
v = list()
w.append(0)
v.append(0)

for i in range(N):
	bags = input().split(" ")
	w.append(int(bags[0]))
	v.append(int(bags[1]))

solution(N, K)
print(dp[N][K])

File: c/b12865.c
#include <stdio.h>

// 12865: 진법 변환

int w[101] = {0,};
int v[101] = {0,};
int dp[101][101] = {0,};

int max(int a, int b) {
	if (a > b) {
		return a;
	}
	return b;
}

void solution(int N, int K) {
	int i, j;
	for (i = 1; i <= K; i++) {
		for (j = 1; j <= N; j++) {
			if (w[j] > i) {
				dp[j][i] = dp[j-1][i];
			} else {
				dp[j][i] = max(dp[j-1][i-w[j]]+v[j], dp[j-1][i]);
			}
		}
	}
}

int main(void) {
	int N, K, i, j;
	for (i = 0; i <= N; i++) {
		for (j = 0; j <= K; j++) {
			dp[i][j] = 0;
		}
	}
	scanf("%d %d", &N, &K);
	for (i = 0; i < N; i++) {
		scanf("%d %d", &w[i+1], &v[i+1]);
	}

	solution(N, K);
	printf("%d", dp[N][K]);

	return 0;
}